tests/zkevm/test_worst_memory.py

```diff
@@ -0,0 +1,335 @@
+
+REFERENCE_SPEC_GIT_PATH = "TODO"
+REFERENCE_SPEC_VERSION = "TODO"
+
+
+class CallDataOrigin:
+    """Enum for calldata origins."""
+
+    TRANSACTION = 1
+    CALL = 2
+
+
+@pytest.mark.valid_from("Cancun")
+@pytest.mark.parametrize(
+    "origin",
+    [
+        pytest.param(CallDataOrigin.TRANSACTION, id="transaction"),
+        pytest.param(CallDataOrigin.CALL, id="call"),
+    ],
+)
+@pytest.mark.parametrize(
+    "size",
+    [
+        pytest.param(0, id="0 bytes"),
+        pytest.param(10, id="10 bytes"),
+        pytest.param(100, id="100 bytes"),
+        pytest.param(1 * 1024, id="1KiB"),
+        pytest.param(10 * 1024, id="10KiB"),
+        pytest.param(100 * 1024, id="100KiB"),
+        pytest.param(1024 * 1024, id="1MiB"),
+    ],
+)
+@pytest.mark.parametrize(
+    "fixed_src_dst",
+    [
+        True,
+        False,
+    ],
+)
+@pytest.mark.parametrize(
+    "non_zero_data",
+    [
+        True,
+        False,
+    ],
+)
+def test_worst_calldatacopy(
+    state_test: StateTestFiller,
+    pre: Alloc,
+    fork: Fork,
+    origin: CallDataOrigin,
+    size: int,
+    fixed_src_dst: bool,
+    non_zero_data: bool,
+):
+    """Test running a block filled with CALLDATACOPY executions."""
+    env = Environment()
+
+    if size == 0 and non_zero_data:
+        pytest.skip("Non-zero data with size 0 is not applicable.")
+
+    # We create the contract that will be doing the CALLDATACOPY multiple times.
+    #
+    # If `non_zero_data` is True, we leverage CALLDATASIZE for the copy length. Otherwise, since we
+    # don't send zero data explicitly via calldata, PUSH the target size and use DUP1 to copy it.
+    prefix = Bytecode() if non_zero_data or size == 0 else Op.PUSH32(size)
+    src_dst = 0 if fixed_src_dst else Op.MOD(Op.GAS, 7)
+    attack_block = Op.CALLDATACOPY(
+        src_dst, src_dst, Op.CALLDATASIZE if non_zero_data or size == 0 else Op.DUP1
+    )
+    code = code_loop_precompile_call(prefix, attack_block, fork)
+    code_address = pre.deploy_contract(code=code)
+
+    tx_target = code_address
+
+    # If the origin is CALL, we need to create a contract that will call the target contract with
+    # the calldata.
+    if origin == CallDataOrigin.CALL:
+        # If `non_zero_data` is False we leverage just using zeroed memory. Otherwise, we
+        # copy the calldata received from the transaction.
+        prefix = (
+            Op.CALLDATACOPY(Op.PUSH0, Op.PUSH0, Op.CALLDATASIZE) if non_zero_data else Bytecode()
+        )
+        arg_size = Op.CALLDATASIZE if non_zero_data else size
+        code = prefix + Op.STATICCALL(
+            address=code_address, args_offset=Op.PUSH0, args_size=arg_size
+        )
+        tx_target = pre.deploy_contract(code=code)
+
+    # If `non_zero_data` is True, we fill the calldata with deterministic random data.
+    # Note that if `size == 0` and `non_zero_data` is a skipped case.
+    data = Bytes([i % 256 for i in range(size)]) if non_zero_data else Bytes()
```

Comment:
I was about to comment that `non_zero_data=False` always yields empty bytes, but this is handled by `size` so we dont "waste" gas on calldata. I like it! :smile: :+1: